src/mscorlib/shared/System/Byte.cs

```diff
@@ -196,6 +196,11 @@ public String ToString(String format, IFormatProvider provider)
             return Number.FormatInt32(m_value, format, NumberFormatInfo.GetInstance(provider));
         }
 
+        public bool TryFormat(Span<char> destination, out int charsWritten, string format = null, IFormatProvider provider = null)
```

Comment:
In UTF8 APIs, we use StandardFormat instead of format string to avoid having to parse the string. Would we want to use StandardFormat here too?

Reply:
How would that support custom formats? I'd have to look at the profiling data, but for these standard formats, IIRC passing the format string was maybe only 2% of the overall cost of the operation.

Reply:
Yeah, you are right. Custom formats is a problem. I think string is ok for now. We can always add optimized overloads if we need them.

Though (FYI), It's probably 2% for the average scenario. But there are scenarios, e.g. formatting sections of a GUID, where you repeatedly have to format with D2 and D4. And the formatted numbers are quite small. When I implement it some time ago using strings, the string parsing was significantly more than 2%.

Reply:
Should `string format = null` be `ReadOnlySpan<char> format = default`?

I've been looking at making use of these new `TryFormat` methods to avoid allocations inside the implementation of `StringBuilder.AppendFormat` (used by `string.Format`/interpolated strings).

Right now, if any of the format items has a specified format, it has to allocate substrings.

https://github.com/dotnet/coreclr/blob/fc8bd03dd1dbeab9b7e51e8ecdce3d3c6d9bdfcc/src/mscorlib/shared/System/Text/StringBuilder.cs#L1510

It'd be nice if those could be non-allocating slices.

Reply:
> But there are scenarios, e.g. formatting sections of a GUID, where you repeatedly have to format with D2 and D4. And the formatted numbers are quite small. When I implement it some time ago using strings, the string parsing was significantly more than 2%.

It is more, though also far from dominating.  I took your GUID formatting example:
```C#
using System;
using System.Diagnostics;
using System.Globalization;

class Program
{
    static void Main()
    {
        uint a = 0x159491E6;
        ushort b = 0x1BC1;
        ushort c = 0xB0D9;
        ulong d = 0x1D477C38F186;

        var sw = new Stopwatch();
        var arr = new char[100];
        int charsWritten;

        var provider = NumberFormatInfo.GetInstance(CultureInfo.CurrentCulture);
        for (int i = 0; i < 50_000_000; i++)
        {
            Span<char> dst = arr;

            a.TryFormat(dst, out charsWritten, "X8", provider);
            dst = dst.Slice(charsWritten);

            b.TryFormat(dst, out charsWritten, "X4", provider);
            dst = dst.Slice(charsWritten);

            c.TryFormat(dst, out charsWritten, "X4", provider);
            dst = dst.Slice(charsWritten);

            d.TryFormat(dst, out charsWritten, "X12", provider);
        }
    }
}
```

Here's a profile:
![image](https://user-images.githubusercontent.com/2642209/32960668-cf0aaa24-cb93-11e7-8fdf-83e70a40f2be.png)

At least in that example:
- `ParseFormatSpecifier` is ~5.3%.  That's not nothing, but it's certainly not the majority contributor.  We might also be able to tweak it a bit to eek out some more throughput.
- `ReadOnlySpan.TryCopyTo` is ~9.9%, but the `Buffer.Memmove` it's using for the actual copy is ~3.9%, so there's 6% in there that going to something other than the actual copy (the trace shows ~2.6% in the `ReadOnlySpan.TryCopyTo` body and ~3.4% in the `Span.CopyTo` body that `TryCopyTo` calls).  If we're looking to optimize for such percentages, I'd prefer to see us start by looking at making TryCopyTo faster, as that'll accrue to many other scenarios.  If we then also want to look at overloads of `TryFormat` that take a `StandardFormat`, we certainly can.  As you say, we can always add them later.

To be clear, I'm not against adding `StandardFormat` to the core and adding overrides that use it; I just don't think that takes the place of the string-based format, and I'm not sure it's currently the most pressing matter.

Of course, it's also possible/likely that there are ways to further optimize methods like `TryInt32ToHexStr`, at which point with that getting faster, the costs associated with the parsing will increase relative to it.

Reply:
ps I realize I'm missing a guid segment in that code :smile:

Reply:
Opened https://github.com/dotnet/coreclr/issues/15076 on the Span.TryCopyTo perf.

Reply:
> Should `string format = null` be `ReadOnlySpan<char> format = default`?

Hmm.  We'd settled on `string` because we didn't see a need for dynamically-generated format strings.  I guess the need does exist.  I wonder if it exists in enough other places to warrant changing the public API; obviously StringBuilder could get away with an internal implementation that supported that.

@KrzysztofCwalina, @terrajobst, any opinions on whether we should change `string format` in all of our new parsing and formatting methods to accept a `ReadOnlySpan<char>` instead of a `string`?  It opens up such possible calls, at the expense of making it slightly harder to use, e.g. the common case of passing in a format string like `"D4"` instead becomes `"D4".AsReadOnlySpan()`, unless we add an implicit cast from `string` to `ReadOnlySpan<char>`.

Reply:
@stephentoub, what is "It opens up such possible calls" referring to? i.e. when would I want to pass span of chars as a format?

Reply:
> when would I want to pass span of chars as a format?

See @justinvp's scenario above.

Reply:
Would StandardFormat overloads address this as well?

Reply:
> Would StandardFormat overloads address this as well?

For standard formats :smile:  AppendFormat supports custom formats, too, though.

Reply:
@KrzysztofCwalina, here's a concrete example:

```c#
DateTime now = DateTime.UtcNow;
Guid id = Guid.NewGuid();
decimal value = 3.50m;

string log = string.Format("{0:s}: Event {1:B} occurred: The value is: {2:C2}", now, id, value);
```

`string.Format`'s implementation has to extract the formats for each format item, e.g. "s", "B", and "C2". It'd be nice if we didn't have to allocate strings for these.

Of course, if this is the only place where this comes up, we could get away with internal support for `ReadOnlySpan<char>`, as @stephentoub mentions.

Reply:
I've opened https://github.com/dotnet/corefx/issues/25337 to track this.  We don't need to hold up this PR for it.